### ISSUE TITLE
Channel adapter: generic supervisor framework with auto-restart (F22)

### ISSUE THREAD

**Issue by suzuke** [suzuke/agend-terminal#695]
## Background
Surfaced by health audit 2026-05 (5-backend dialectic, raised by auditor-opencode as I-9, AGREE 5/5).

`src/channel/telegram/inbound.rs:13-48` spawns the Telegram polling dispatcher with **no JoinHandle, no health check, no restart-on-failure mechanism**:

```rust
// fire-and-forget: Stops when the bot's update stream errors
// (network drop / shutdown). No JoinHandle / shutdown signal needed —
// process exit reaps the thread.
std::thread::spawn(move || {
    teloxide::Dispatcher::builder(...)
        .build()
        .dispatch(); // ← any error here, thread exits permanently
});
```

= Telegram connectivity is **silently lost** the moment Dispatcher returns (network drop, token revoked, API rate limit, internal panic). Daemon stays alive but loses the channel; operator notices only when their next telegram message doesn't reach an agent — potentially 30 min to several hours of silent failure.

## Operator decision (2026-05-12, option b)

Generic supervisor framework for **all channel adapters**, not just telegram. Discord (scaffold-only today) automatically inherits when production-ready. Future channels (Slack / IRC / etc) just implement the trait.

### Channel adapter trait extension

```rust
pub trait ChannelAdapter: Send + Sync {
    fn name(&self) -> &'static str;

    /// Spawn the channel's main poll/dispatch loop. Caller (supervisor)
    /// holds the JoinHandle.
    fn spawn_loop(&self) -> Result<JoinHandle<ChannelExitReason>>;

    /// Backoff schedule for transient errors. Default exponential.
    fn backoff_schedule(&self) -> &[Duration] { &DEFAULT_BACKOFF }

    /// Max consecutive failures before fatal escalation.
    fn max_failures(&self) -> u32 { 5 }

    /// Channel-specific health probe (does the channel still respond?).
    /// Called periodically by supervisor between restarts.
    fn health_check(&self) -> Result<HealthState>;
}

pub enum ChannelExitReason {
    Transient(String),     // network drop, etc — supervisor retries
    Fatal(String),         // token revoked, auth gone — supervisor escalates
    GracefulShutdown,      // daemon stopping — supervisor exits
}
```

### Supervisor

`src/channel/supervisor.rs` (new file ~100 LOC):
- Spawns adapter's loop
- On exit: classify reason, apply backoff, restart up to `max_failures`
- On fatal / max-failures exceeded: emit `[channel_fatal]` event to all OTHER live channels + log
- Periodic health_check between activity, force restart if degraded

### Cross-channel fatal alert

When Telegram supervisor exits fatally, route alert via:
1. Any other live channel (Discord when ready, etc)
2. Persistent log file `~/.agend-terminal/channel_alerts.log`
3. (future) systemd journal entry / desktop notification

When Discord supervisor exits fatally, route alert via Telegram (mirrors).

## Implementation phasing

**PR-1: Trait + supervisor framework** (~80-120 LOC, no behavioral change)
- Add `ChannelAdapter` trait extension
- Add `Supervisor` struct
- Wire supervisor into daemon main loop registry
- All current `fire-and-forget` channel spawns remain in place (no semantic change yet)

**PR-2: Telegram migrates to supervisor** (~60-80 LOC)
- Telegram `inbound.rs` implements `ChannelAdapter` trait
- Old `std::thread::spawn` site replaced with `Supervisor::adopt(telegram_adapter)`
- Telegram-specific `backoff_schedule` (rate-limit-aware) + `max_failures`
- Telegram `health_check` (e.g. `getMe` API call)

**PR-3: Discord scaffold conforms** (~30-50 LOC)
- Discord adapter implements trait shape (impl returns Err(NotSupported) where appropriate)
- Discord scaffold ready for future production-ready PR

Total ~150-250 LOC across 3 PRs.

## Test required

- Unit test: supervisor classification of Transient vs Fatal vs GracefulShutdown
- Integration test (telegram): mock dispatcher returns Transient error → supervisor restarts after backoff → resumes normal flow
- Integration test (telegram): mock dispatcher returns Fatal → supervisor exits + fires cross-channel alert
- Integration test: max_failures exceeded → escalation path triggered

## Evidence

- `src/channel/telegram/inbound.rs:13-48` — current fire-and-forget pattern + rationale comment
- 5-backend cross-verified AGREE 5/5

## Severity

Important. Telegram is the operator's primary out-of-TUI channel. Silent connectivity loss = operator surprise + delayed fleet awareness.

## Related
- F4 (skipped) — telegram topic-close authz, different concern same channel layer
- F11 (#686) — coverage tooling, helps validate refactor doesn't regress
- F20 (#694) — daemon event loop refactor, supervisor is one such subsystem-extraction sub-case

**Comment by suzuke** [comment]
**Decision: P4 backlog — not fixing now.**

Rationale: Over-abstraction. Only one channel adapter (telegram) exists; no second channel demand. Will revisit when multi-channel support is needed.

**Comment by suzuke** [comment]
## Audit team Phase 4 review (5-backend dialectic)

5 個 auditor 對 dev-team「over-abstraction、no second channel」P4 defer rationale 的表決：

| Auditor | 立場 |
|---|---|
| auditor-codex | DISAGREE — 「目前只有 Telegram」不構成 defer 理由、單通道失效是現有可用性 hole |
| auditor-gemini | DISAGREE — 「no second channel」strawman、Telegram polling thread fragile + 任何 API/network 失敗就死 |
| auditor-kiro | DISAGREE — 應該 P2、~20 LOC 簡單 retry loop 即可、不需要 framework |
| auditor-opencode | DISAGREE — 混淆「supervisor framework」(YAGNI) 跟「single-channel restart hardening」(real reliability gap) |
| auditor-claude | DISAGREE — 切兩個 scope：framework P4 維持、telegram restart-hardening 立即 P2 narrow fix |

**共識**：5/5 DISAGREE。dev-team 把兩個不同 scope 混在一起：
1. **Channel adapter framework 抽象**（多 channel 用） → 同意 YAGNI、P4 OK
2. **Telegram polling thread restart-on-failure**（單 channel reliability hardening） → 跟 multi-channel 無關、是現有 production 路徑風險

**Audit team 建議**：split issue scope。framework 部分維持 P4。Telegram 自身 restart hardening 立刻 P2、~30 LOC `loop { dispatch().await; warn!(\"restart\"); sleep(backoff) }` retry wrapper、無 framework dependency。
